packages/shared/src/effect.ts

```diff
@@ -31,68 +31,67 @@ export const fetchEff = (
   input: RequestInfo | URL,
   init?: RequestInit,
 ): Effect.Effect<ResponseEsque, FetchError, FetchContext> =>
-  FetchContext.pipe(
-    Effect.andThen(({ fetch, baseHeaders }) =>
-      Effect.tryPromise({
-        try: () =>
-          fetch(input, {
-            ...init,
-            headers: {
-              ...filterObjectValues(baseHeaders, (v): v is string => v != null),
-              ...init?.headers,
-            },
-          }),
-        catch: (error) => new FetchError({ error, input }),
+  Effect.flatMap(FetchContext, ({ fetch, baseHeaders }) => {
+    const req = new Request(input, {
+      ...init,
+      headers: {
+        ...filterObjectValues(baseHeaders, (v): v is string => v != null),
+        ...init?.headers,
+      },
+    });
+    return Effect.tryPromise({
+      try: () => fetch(req.clone()),
+      catch: (error) => new FetchError({ error, input: req.clone() }),
+    }).pipe(
+      Effect.map((res) => Object.assign(res.clone(), { request: req.clone() })),
```

Comment:
I'm just not sure when the `Request` is actually consumed and couldn't be bothered to check the docs hahahaha
I was crossing my fingers you'll know when clone is required and when its not.

Assigning the request to the response is because of a business requirement uploadthing establishes that responses need to know about their requests. Which is totally valid but I wasn't sure you want this requirement to stretch your scope.. With this you can access the request wherever you need to as long as you have access to the response.

I'll fix the clone, and I'm open to ideas about the assignment